src/ClientData/include/ClientData/ModuleData.h

```diff
@@ -47,27 +46,39 @@ class ModuleData final {
   [[nodiscard]] uint64_t ConvertFromVirtualAddressToOffsetInFile(uint64_t virtual_address) const;
   [[nodiscard]] uint64_t ConvertFromOffsetInFileToVirtualAddress(uint64_t offset_in_file) const;
 
-  [[nodiscard]] bool is_loaded() const;
-  // Returns true if the module was unloaded and false otherwise.
+  // Returns true if the module was unloaded (symbols were removed) and false otherwise.
   [[nodiscard]] bool UpdateIfChangedAndUnload(orbit_grpc_protos::ModuleInfo info);
-  // This method does not update module_data in case it needs update and was not loaded.
-  // returns true if update was successful or no update was needed and false if module
-  // cannot be updated because it was loaded.
+  // This method does not update the module in case symbols are already loaded, even if the module
+  // would need to be updated. Returns true if the update was successful or no update was needed,
+  // and false if the module cannot be updated because symbols are already loaded.
   [[nodiscard]] bool UpdateIfChangedAndNotLoaded(orbit_grpc_protos::ModuleInfo info);
 
   [[nodiscard]] const FunctionInfo* FindFunctionByVirtualAddress(uint64_t virtual_address,
                                                                  bool is_exact) const;
-  void AddSymbols(const orbit_grpc_protos::ModuleSymbols& module_symbols);
   [[nodiscard]] const FunctionInfo* FindFunctionFromHash(uint64_t hash) const;
   [[nodiscard]] const FunctionInfo* FindFunctionFromPrettyName(std::string_view pretty_name) const;
   [[nodiscard]] std::vector<const FunctionInfo*> GetFunctions() const;
 
+  enum class SymbolCompleteness {
+    kNoSymbols = 0,
+    kDynamicLinkingAndUnwindInfo = 1,  // i.e., "fallback symbols"
```

Comment:
I am wondering if there should be two cases: 1 dynamic linking info and 2 unwind info?

Reply:
I was wondering about the following case (unlikely to happen though):
* Object file does not contain any symbols (no dynsym in particular), but unwind info
* We have a debug file that only contains dynsym
* The debug file is first not found by the symbol paths

Will we be able to "update" the symbols if we make the debug file available?

Reply:
It might be possible to fabricate the situation you describe by manipulating the file somehow, but it's not a valid one and we don't support it: .dynsym is used for dynamic linking, it has to be in the binary; conversely, a debug file that only contains .dynsym is not a debug file.

Reply:
It might be possible to fabricate the situation you describe by manipulating the file somehow, but it's not a valid one and we don't support it: .dynsym is used for dynamic linking, it has to be in the binary; conversely, a debug file that only contains .dynsym is not a debug file.